addons/mass_mailing/models/mailing_list.py

```diff
@@ -222,41 +246,73 @@ def action_merge(self, src_lists, archive):
         self.ensure_one()
         # Put destination is sources lists if not already the case
         src_lists |= self
-        self.env.flush_all()
-        self.env.cr.execute("""
+        self.env['mailing.contact'].flush_model()
+        contact_condition = self._mailing_list_get_contact_condition()
+        select_values = self._mailing_list_get_select_values()
+        rn = self._mailing_list_get_rn()
+        rn_where = self._mailing_list_get_rn_where()
+        select = ', '.join("COALESCE(%s, '')" % field for field in select_values.split(', '))
+        join_select = ', '.join("COALESCE(%s, '')" % ("sub_" + field) for field in select_values.split(', ')) #Need to add coalesce as in postgresql (NULL, 1) = 1
+        insert_query = f"""
             INSERT INTO mailing_contact_list_rel (contact_id, list_id)
-            SELECT st.contact_id AS contact_id, %s AS list_id
-            FROM
-                (
-                SELECT
-                    contact.id AS contact_id,
-                    contact.email AS email,
-                    list.id AS list_id,
-                    row_number() OVER (PARTITION BY email ORDER BY email) AS rn
-                FROM
-                    mailing_contact contact,
-                    mailing_contact_list_rel contact_list_rel,
-                    mailing_list list
+            SELECT contact_id, %(id)s AS list_id
+            FROM (
+                SELECT contact.id AS contact_id,
+                       list.id AS list_id,
+                       {select_values},
```

Comment:
so here you will need to join
```suggestion
                       {','.join(select_values)},
```